source/common/event/scaled_timer.cc

```diff
@@ -0,0 +1,207 @@
+
+void ScaledTimerManager::ScaledTimerList::addActive(ScaledTimerImpl* timer) {
+  ASSERT(timer->active().has_value());
+  timers_.insert(timer);
+}
+
+void ScaledTimerManager::ScaledTimerList::removeActive(ScaledTimerImpl* timer) {
+  ASSERT(timer->active().has_value());
+  timers_.erase(timer);
+}
+
+void ScaledTimerManager::ScaledTimerList::setScaleFactor(double scale_factor) {
+  std::set<ScaledTimerImpl*, Compare> reordered(timers_.begin(), timers_.end(),
+                                                Compare{.scale_factor = scale_factor});
+  std::swap(reordered, timers_);
+}
+
+absl::optional<ScaledTimerManager::ScaledTimerImpl*>
+ScaledTimerManager::ScaledTimerList::pop(MonotonicTime now) {
+  if (timers_.empty()) {
+    return absl::nullopt;
+  }
+
+  auto* timer = *timers_.begin();
+  if (timers_.key_comp().targetTime(timer) <= now) {
+    timers_.erase(timers_.begin());
+    return timer;
+  }
+  return absl::nullopt;
+}
+
+absl::optional<MonotonicTime> ScaledTimerManager::ScaledTimerList::firstScheduleTime() const {
+  if (timers_.empty()) {
+    return absl::nullopt;
+  }
+  return timers_.key_comp().targetTime(*timers_.begin());
+}
+
+MonotonicTime
+ScaledTimerManager::ScaledTimerList::Compare::targetTime(const ScaledTimerImpl* timer) const {
+  ASSERT(timer->active().has_value());
+
+  auto scaled_delay = timer->active()->delay * scale_factor;
+  return timer->active()->start +
+         std::chrono::duration_cast<std::chrono::microseconds>(scaled_delay);
+}
+
+bool ScaledTimerManager::ScaledTimerList::Compare::operator()(const ScaledTimerImpl* lhs,
+                                                              const ScaledTimerImpl* rhs) const {
+  return targetTime(lhs) < targetTime(rhs);
+}
+
+void ScaledTimerManager::scheduleTimer(ScaledTimerImpl* timer) {
+  ASSERT(timer->active().has_value());
+
+  timers_.addActive(timer);
+  internalScheduleTimer();
+}
+
+void ScaledTimerManager::descheduleTimer(ScaledTimerImpl* timer) {
+  ASSERT(timer->active().has_value());
+  timers_.removeActive(timer);
+  internalScheduleTimer();
+}
+
+void ScaledTimerManager::dispatchCallback() {
+  // Find all timers that are ready to be handled and dispatch them.
+  const auto now = dispatcher_.timeSource().monotonicTime();
+  for (auto timer = timers_.pop(now); timer.has_value(); timer = timers_.pop(now)) {
+    timer.value()->dispatch(dispatcher_);
+  }
+
+  internalScheduleTimer();
+}
+
+void ScaledTimerManager::internalScheduleTimer() {
+  auto next_time = timers_.firstScheduleTime();
+  if (!next_time.has_value()) {
+    dispatch_timer_->disableTimer();
+    return;
+  }
+  const MonotonicTime now = dispatcher_.timeSource().monotonicTime();
+  const auto target_delay = std::chrono::duration_cast<std::chrono::microseconds>(*next_time - now);
+
+  if (target_delay > std::chrono::microseconds::zero()) {
+    dispatch_timer_->enableHRTimer(target_delay);
+  } else {
+    dispatch_timer_->enableHRTimer(std::chrono::microseconds::zero());
```

Comment:
I have an upcoming change that will break this since 0msec timer scheduling will no longer happen in the same iteration as this call.  But the breakage is more fundamental: 0ms timers scheduled via this interface instead of the regular one would execute in the current iteration instead of the next iteration due to the way they are merged with other timers.  I need to think of a way to achieve the right semantics; it may require keeping new additions separate and use of an update callback to move the newly added callbacks to the main set and adjust dispatch_timer_